--- qemu-kvm.spec.orig	2016-06-03 11:32:38.318001856 +0300
+++ qemu-kvm.spec	2016-06-03 11:37:19.059000486 +0300
@@ -104,7 +104,7 @@
 # that the kernel isn't the stock distribution qemu-kvm, for example,
 # by setting the define to ".local" or ".bz123456"
 
-%define buildid %{nil}
+%define buildid .onapp.1
 
 %define zrelease 4
 %define sublevel 0.12.1.2
@@ -13386,6 +13386,7 @@ cd qemu-kvm-x86_64-build
             --enable-lzo \
             --enable-snappy \
             --enable-vhdx \
+            --enable-io-throttling \
             %{config_fake_machine} \
             %{disable_rhev_features_arg}
 
@@ -13666,6 +13667,9 @@ fi
 %endif # with qemu_kvm
 
 %changelog
+* Fri Jun 3 2016 Yuriy Kohut <ykohut@onapp.com> - 0.12.1.2-2.479.el6_7.4.onapp.1
+- configure if qemu-kvm with --enable-io-throttling
+
 * Mon Jan 25 2016 Jeff E. Nelson <jen@redhat.com> - 0.12.1.2-2.479.el6_7.4
 - kvm-fw_cfg-add-check-to-validate-current-entry-value-CVE.patch [bz#1298045]
 - Resolves: bz#1298045
